arangod/Aql/AsyncPrefetchSlotsManager.cpp

```diff
@@ -0,0 +1,114 @@
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "AsyncPrefetchSlotsManager.h"
+#include "Assertions/Assert.h"
+
+#include <algorithm>
+#include <utility>
+
+namespace arangodb::aql {
+
+AsyncPrefetchSlotsReservation::AsyncPrefetchSlotsReservation() noexcept
+    : _manager(nullptr), _slotsReserved(0) {}
+
+AsyncPrefetchSlotsReservation::AsyncPrefetchSlotsReservation(
+    AsyncPrefetchSlotsManager* manager, size_t slotsReserved) noexcept
+    : _manager(manager), _slotsReserved(slotsReserved) {
+  // if number of slots is > 0, we must have a manager
+  TRI_ASSERT(_slotsReserved == 0 || _manager != nullptr);
+}
+
+AsyncPrefetchSlotsReservation& AsyncPrefetchSlotsReservation::operator=(
+    AsyncPrefetchSlotsReservation&& other) noexcept {
+  if (this != &other) {
+    returnSlots();
+    if (other._manager != nullptr) {
+      _manager = other._manager;
+    }
+    _slotsReserved = std::exchange(other._slotsReserved, 0);
+  }
+  return *this;
+}
+
+AsyncPrefetchSlotsReservation::~AsyncPrefetchSlotsReservation() {
+  returnSlots();
+}
+
+void AsyncPrefetchSlotsReservation::returnSlots() noexcept {
+  if (_manager != nullptr && _slotsReserved > 0) {
+    _manager->returnSlots(_slotsReserved);
+    _slotsReserved = 0;
+    // leave _manager in place
+  }
+  TRI_ASSERT(_slotsReserved == 0);
+}
+
+AsyncPrefetchSlotsManager::AsyncPrefetchSlotsManager() noexcept
+    : AsyncPrefetchSlotsManager(0, 0) {}
+
+AsyncPrefetchSlotsManager::AsyncPrefetchSlotsManager(
+    size_t maxSlotsTotal, size_t maxSlotsPerQuery) noexcept
+    : _maxSlotsTotal(maxSlotsTotal),
+      _maxSlotsPerQuery(maxSlotsPerQuery),
+      _slotsUsed(0) {}
+
+void AsyncPrefetchSlotsManager::configure(size_t maxSlotsTotal,
+                                          size_t maxSlotsPerQuery) noexcept {
+  _maxSlotsTotal = maxSlotsTotal;
+  _maxSlotsPerQuery = maxSlotsPerQuery;
+}
+
+AsyncPrefetchSlotsReservation AsyncPrefetchSlotsManager::leaseSlots(
+    size_t value) noexcept {
+  // reduce value to the allowed per-query maximum
+  value = std::min(_maxSlotsPerQuery, value);
+
+  if (value > 0) {
+    // check if global (per-server) total is below the configured total maximum
+    size_t expected = _slotsUsed.load(std::memory_order_relaxed);
+    TRI_ASSERT(expected <= _maxSlotsTotal);
+    // as long as we have slots left, try to allocate some
+    while (expected < _maxSlotsTotal) {
+      size_t target = std::min(_maxSlotsTotal, expected + value);
+      TRI_ASSERT(target > expected);
+      if (_slotsUsed.compare_exchange_weak(expected, target,
+                                           std::memory_order_relaxed)) {
+        return {this, target - expected};
+      }
+      TRI_ASSERT(expected <= _maxSlotsTotal);
+    }
```

Comment:
Maybe we should rename `expected` to `current`. IMHO this makes the code clearer and easier to understand
```suggestion
    size_t current= _slotsUsed.load(std::memory_order_relaxed);
    TRI_ASSERT(current <= _maxSlotsTotal);
    // as long as we have slots left, try to allocate some
    while (current < _maxSlotsTotal) {
      size_t target = std::min(_maxSlotsTotal, current + value);
      TRI_ASSERT(target > current);
      if (_slotsUsed.compare_exchange_weak(current, target,
                                           std::memory_order_relaxed)) {
        return {this, target - current};
      }
      TRI_ASSERT(current <= _maxSlotsTotal);
    }
```